src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.EasyRequest.cs

```diff
@@ -249,10 +249,51 @@ private void SetUrl()
                 }
 
                 EventSourceTrace("Url: {0}", requestUri);
-                SetCurlOption(CURLoption.CURLOPT_URL, requestUri.AbsoluteUri);
+                if (requestUri.Host != requestUri.IdnHost)
+                {
+                    SetCurlOption(CURLoption.CURLOPT_URL, ReplaceHostWithIdnHost(requestUri.AbsoluteUri, requestUri.Host, requestUri.IdnHost));
+                }
+                else
+                {
+                    SetCurlOption(CURLoption.CURLOPT_URL, requestUri.AbsoluteUri);
+                }
+
                 SetCurlOption(CURLoption.CURLOPT_PROTOCOLS, (long)(CurlProtocols.CURLPROTO_HTTP | CurlProtocols.CURLPROTO_HTTPS));
             }
 
+            private unsafe string ReplaceHostWithIdnHost(string absoluteUri, string host, string idnHost)
+            {
+                int start = absoluteUri.IndexOf(host);
```

Comment:
I would also prefer UriBuilder, since uri syntax parsing will be kept in Uri apis, and we will not have to keep this parsing in HttpClient in sync on future syntax updates.